src/bit_array.cr

```diff
@@ -32,10 +32,17 @@ struct BitArray
 
   def ==(other : BitArray)
     return false if size != other.size
-    # NOTE: If BitArray implements resizing, there may be more than 1 binary
-    # representation and their hashes for equivalent BitArrays after a downsize as the
-    # discarded bits may not have been zeroed.
-    return LibC.memcmp(@bits, other.@bits, malloc_size) == 0
+    return true if size == 0
```

Comment:
Could you explain why this entire change is needed? I don't understand why the last bit needs special handling. And if it does, there should be a comment in the code explaining why.

Reply:
It needs special handling since unused bits at the end of the array are not guaranteed to be in a defined state, so they have to be masked out (i.e., set to zero) before comparing.

Reply:
No, that's not correct. `Pointer.malloc` makes sure everything is zero. The comparison is broken because `LibC.memcmp` expects the number of bytes but it uses `malloc_size` which is in terms of UInt32. Changing it to use `malloc_size * sizeof(UInt32)` makes it work.

Reply:
Oh, I see. The method `[]` will copy bits from an existing bit array and just copy them over without zeroing the non-interesting part, that's why `memcmp` doesn't work.

I think we should fix `[]` to guarantee the invariant that unused bits are always zero.

Reply:
I don't really care which method is used, this PR is certainly fine.